rfcs/xxxx-collage.md

```diff
@@ -0,0 +1,987 @@
+}
+```
+
+Ideally this optimal partitioning would be understandable to the user, see Appendix G.
+
+## Design
+
+The implementation is mostly under `src/relay/collage/...` (namespace `tvm::relay::collage`), with just a few Python
+helper functions under `python/tvm/relay/collage`.
+
+If the `relay.collage.enable_collage` `PassConfig` attribute is true then a new `CollagePartitioner` pass is inserted
+before all other Relay passes. The result of the pass is:
+
+- All Relay sub-graphs in all global functions which are to be handed off to a BYOC toolchain are replaced by calls to
+  an inline `"Primitive"` function with `"Compiler"` and `"global_symbol"` attributes.
+- Relay operators, or groups of operators, which are to be translated to particular library or BYOC-supplied function
+  are replaced by calls to an inline `"Composite"` function. (This encoding is supported for both BYOC and external
+  libraries.)
+
+Note that no `"Primitive"` functions denoting TVM kernels are produced -- the existing `FuseOps` pass is still required.
+
+The `CollagePartitioner` pass has four phases:
+
+- **Phase 1**: The available `Target`s are scanned to build a list of rules describing how to find possible partitions (
+  see `PartitionSpec` and `PartitionRule` below). Depending on the `Target` the rules may incorporate entries from the
+  BYOC pattern table. (The remaining phases execute on each global function separately.)
+- **Phase 2**: A dataflow graph is constructed for the global function (which is just an `IndexedGraph<Expr>`). The
+  available rules from phase 1 are evaluated on the dataflow graph to yield a (possibly overlapping) set of candidate
+  partitions for each target (see `CandidatePartition` below). Each candidate efficiently describes a sub-graph of the
+  global function's body without the need to construct any new expressions (see `SubGraph` below).
+- **Phase 3**: A shortest path is found in the following (implicit and lazily constructed) search graph:
+    - Search Nodes: The set of dataflow nodes which have already been assigned to a candidate partition in all paths to
+      the node.
+    - Search Edge X->Y: A candidate partition can be applied to node X to give node Y. The candidate is disjoint from
+      all dataflow nodes already assigned in X. To avoid an unnecessary search space explosion the candidate must also
+      include the next yet-to-be-assigned dataflow node in X.
+    - Edge cost: The estimated latency of the candidate partition, plus a partition transition penalty. Note that though
+      we need to be able to extract the candidate's sub-graph in order to build a function representing the candidate,
+      we do not yet need to partition the overall function body expression.
+
+  Other search algorithms are certainly possible, eg the paper uses an evolutionary search to refine the partitioning
+  found by the dynamic-programming search. We can easily abstract away the search interface to support multiple
```

Comment:
If this is an aspiration, we need to define a bit higher-level interface where path-finding formulation becomes a specific implementation.

Reply:
I think that's best deferred till we have the second example. At this point however I think IRModule -> Array<CandidatePartition> looks to be about as specific as we can get.